beam-sync.md

```diff
@@ -0,0 +1,187 @@
+
+#### Optimization 2: Parallel Execution
+
+The flow diagram above shows block imports as serial, so that child blocks aren't executed until the parent is complete. We want to improve on that with Parallel Execution. We start executing child blocks before the parent has completed executing. Some implications:
+
+- The JIT-data limit of only requesting one trie node at a time applies per-block, so they can be requested in parallel across all pending blocks.
+- The block count lag determines how many pending blocks are available. If you are right at the tip, parallelism drops to 1. The benefit of previewing grows as you lag and shrinks as you catch up. It is great to catch up faster, the further you are behind, but it can make Beam Sync feel a bit "stutter-y".
+- Parallel execution "unnecessarily" retrieves state for child blocks, if the parent block writes that state. The data would have been available locally, if the client had waited for the parent to finish. This adds strain to the network. In practice, though, many blocks are largely independent in state access, so the waste is small.
+
+#### Optimization 3: Speculative Execution
+
+Many transactions included in the same block access independent parts of the state. So we can speculatively execute the transactions out-of-order. To speculatively execute, we treat each transaction as if it was the first & only transaction in the block. The transaction may fail, or execute on a different code path, in which case we get no benefit. But for the many state-independent transactions, we can correctly collect the required JIT-data for each transaction in parallel.
+
+Note: if the same block includes multiple transactions from the same sender, then we group them together to run in serial. If you don't group them together, then the 2nd to nth transaction from the sender would always fail the nonce check, and provide no benefit.
+
+There are probably other games you could play to predict transactions that depend on each other, and serialize them. But it hasn't been explored yet.
+
+## Network Health
+
+There are a few benefits to the health of the network.
+
+One was already mentioned: Beam Syncing peers can serve data to other Beam Syncing peers. This means that a giant swarm of newly-connected nodes do not choke each other from joining the network.
+
+Additionally, there is the cache-friendliness of the design. Storage I/O dominates the cost of serving trie nodes. Since peers are serving the same data that they just received, it should typically be in cache, whether that's at the application level, DB level, or OS level. In fact, since the trie nodes were already accessed for blocks that were recently executed, nodes that are unaware of Beam Sync might even have these trie nodes cached. This reduces the strain of serving state data.
+
+(Fast Sync, by comparison, seems to be purpose-built to wreck the cache, by forcing a node to do a complete read out of all trie data from the database)
```

Comment:
I'd remove the "purpose built" since it almost definitely wasn't (but I understand the sentiment).  Aim for more neutral language.